File: Serialize and Deserialize Binary Tree/Serialize_and_Deserialize_Binary_Tree.cpp
// 二叉树的序列化与反序列化，使用前序遍历解决
// https://leetcode.com/problems/serialize-and-deserialize-binary-tree/discuss/74259/Recursive-preorder-Python-and-C%2B%2B-O(n)
// Runtime: 36 ms, faster than 97.70% of C++ online submissions for Serialize and Deserialize Binary Tree.
// Memory Usage: 26.3 MB, less than 58.80% of C++ online submissions for Serialize and Deserialize Binary Tree.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec 
{
public:

    // Encodes a tree to a single string.
    // 节点之间使用，分隔，空节点使用#进行表示
    string serialize(TreeNode* root) 
    {
        ostringstream os;
        stack<TreeNode*> myStack;

        while (root || !myStack.empty())
        {
            while (root)
            {
                os << root->val << ' ';
                myStack.push(root);
                root = root->left;
            }

            // 需要使用空格进行分割
            os << "# ";

            if (!myStack.empty())
            {
                root = myStack.top();
                myStack.pop();
                root = root->right;
            }
        }

        os << "# ";
        return os.str();
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) 
    {
        istringstream is(data);
        return helper(is);
    }

private:
    TreeNode* helper(istringstream& is)
    {
        // istringstream 永远不会弹完，只会在最后一个string上循环
        string temp;
        is >> temp;

        if (temp == "#")
            return nullptr;

        int value = stoi(temp);
        TreeNode* newNode = new TreeNode(value);

        newNode->left = helper(is);
        newNode->right = helper(is);

        return newNode;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.deserialize(codec.serialize(root));

File: Serialize and Deserialize Binary Tree/Serialize_and_Deserialize_Binary_Tree.py
# Runtime: 120 ms, faster than 75.17% of Python3 online submissions for Serialize and Deserialize Binary Tree.
# Memory Usage: 17.8 MB, less than 14.13% of Python3 online submissions for Serialize and Deserialize Binary Tree.

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:
    
    def __init__(self):
        self.index = 0

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        res = []

        stack = []
        while root or stack:
            while root:
                res.append(str(root.val))
                stack.append(root)
                root = root.left
            res.append('#')

            if stack:
                root = stack.pop()
                root = root.right
        
        res.append('#')
        return ','.join(res)

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        node = data.split(',')
        return self.helper(node, self.index)
    
    def helper(self, node, index):
        if node[index] == '#':
            self.index += 1
            return None
    
        newnode = TreeNode(int(node[index]))
        self.index += 1
        newnode.left = self.helper(node, self.index)
        newnode.right = self.helper(node, self.index)

        return newnode

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))